pkgs/build-support/nextcloud/build-nextcloud-app.nix

```diff
@@ -0,0 +1,179 @@
+
+          patchPhase = ''
+            runHook prePatch
+
+            rm -f Makefile
+
+            runHook postPatch
+          '';
+
+          postPatch = npmPostPatch;
+
+          makeCacheWritable = true;
+          PUPPETEER_SKIP_DOWNLOAD = 1;
+          CYPRESS_INSTALL_BINARY = 0;
+          nativeBuildInputs = npmNativeBuildInputs;
+          buildInputs = npmBuildInputs;
+          preBuild = npmPreBuild;
+
+          installPhase = ''
+            mkdir $out
+            if [ -d js ]; then
+              cp -r js $out
+            fi
+            if [ -d css ]; then
+              cp -r css $out
+            fi
+          '';
+        }
+        // lib.optionalAttrs (npmDeps != null) { inherit npmDeps; }
+        // lib.optionalAttrs (npmDepsHash != null) { inherit npmDepsHash; }
+      )
+    else if builtins.pathExists "${src}/package.json" then
+      throw "package.json exists, but npmDepsHash was not provided"
+    else if builtins.pathExists "${src}/js" || builtins.pathExists "${src}/css" then
+      # An app might ship some Javascript/CSS without npm, but then these are the source files and not compiled ones.
+      stdenvNoCC.mkDerivation {
+        inherit version src;
+        pname = "${pname}-direct";
+
+        patchPhase = ''
+          rm -f Makefile
+        '';
+
+        installPhase = ''
+          mkdir $out
+          if [ -d js ]; then
+            cp -r js $out
+          fi
+          if [ -d css ]; then
+            cp -r css $out
+          fi
+        '';
+      }
+    else
+      null;
+  composerAssets =
+    if vendorHash != null then
+      let
+        assets = php.buildComposerProject2 (finalAttrs: {
+          inherit
+            version
+            src
+            vendorHash
+            ;
+          pname = "${pname}-composer";
+
+          patchPhase = ''
+            runHook prePatch
+
+            rm -f Makefile
+
+            runHook postPatch
+          '';
+
+          composerNoDev = true;
+          composerNoPlugins = true;
+          composerNoScripts = true;
+          composerStrictValidation = false;
+          strictDeps = true;
+
+          postBuild = ''
+            rm -rf vendor/bin
+          '';
+
+          installPhase = ''
+            mkdir $out
+            declare -g composerVendor
+            cp -r "''${composerVendor}/." $out
+          '';
+        });
+        installed_packages =
+          (builtins.fromJSON (builtins.readFile "${assets}/vendor/composer/installed.json")).packages;
+      in
+      if
+        # Just ignore the assets, if no dependencies were installed (e.g. only dev dependencies declared in composer.json).
+        # bamarni/composer-bin-plugin is basically a dev dependency and doesn't need to be available at runtime, but must be installed as a real dependency, so it can be filtered away.
+        builtins.length (
+          builtins.filter (package: package.name != "bamarni/composer-bin-plugin") installed_packages
```

Comment:
Are apps that don't need composer expected to only require this? Wouldn't it make more sense to check if only require-dev is used in composer.json or something like that?